### PR TITLE
Add glue for readablestream external underlying source

### PR DESCRIPTION
**Pull request by gterzian**
Companion to https://github.com/servo/servo/pull/25873

### CODE REVIEW

**Comment by gterzian** [comment]
@asajeffrey r?

Should I increment the version number?

**Comment by jdm** [comment]
Incrementing the version number isn't really necessary since we can't publish to crates.io.

File: src/jsglue.cpp [hunk 1]
```diff
@@ -76,6 +77,52 @@ class RustJobQueue : public JS::JobQueue{
   }
 };
 
+struct ReadableStreamUnderlyingSourceTraps {
+  void (*requestData)(void* source, JSContext* cx, JS::HandleObject stream, size_t desiredSize);
+  void (*writeIntoReadRequestBuffer)(void* source, JSContext* cx, JS::HandleObject stream, void* buffer, size_t length, size_t* bytesWritten);
+  JS::Value (*cancel)(void* source, JSContext* cx, JS::HandleObject stream, JS::HandleValue reason);
```
[the remark this answers is not in the document]
Comment (gterzian) [reply src/jsglue.cpp]: The rust function actually returns a `*mut JS::Value`, and it's the C++ one that returns the value(it seems to work in practice). Actually I tried changing it and the C++ virtual class really wants the overidden function to return a `JS::Value`.

File: src/jsglue.cpp [hunk 2]
```diff
@@ -1063,6 +1110,12 @@ CreateJobQueue(const JobQueueTraps* aTraps, void* aQueue)
   return new RustJobQueue(*aTraps, aQueue);
 }
 
+JS::ReadableStreamUnderlyingSource*
+CreateReadableStreamUnderlyingSource(const ReadableStreamUnderlyingSourceTraps* aTraps, void* aSource)
```
[the remark this answers is not in the document]
Comment (gterzian) [reply src/jsglue.cpp]: Ok I've integrated it with the `finalizer` function, from a manual test running a fetch test, it seems to work in the sense that the finalizer is called, and it doesn't crash.

File: src/glue.rs [hunk 3]
```diff
@@ -20,6 +20,27 @@ impl ::std::default::Default for JobQueueTraps {
...
+pub struct ReadableStreamUnderlyingSourceTraps {
+    pub requestData: ::std::option::Option<
+            unsafe extern "C" fn(source: *const c_void, cx: *mut JSContext, stream: HandleObject,
+                                 desiredSize: usize)>,
+    pub writeIntoReadRequestBuffer: ::std::option::Option<
+            unsafe extern "C" fn(source: *const c_void, cx: *mut JSContext, stream: HandleObject,
+                                 buffer: *mut c_void, length: usize, bytesWritten: *mut usize)>,
+    pub cancel: ::std::option::Option<
+            unsafe extern "C" fn(source: *const c_void, cx: *mut JSContext, stream: HandleObject,
+                                 reason: HandleValue) -> *mut JS::Value>,
```
[the remark this answers is not in the document]
Comment (gterzian) [reply src/glue.rs]: Ok, I've changed it to https://github.com/servo/rust-mozjs/pull/495/files#diff-0c2f4ced2d2186c7a57c783c523a42c1R108

**Comment by gterzian** [comment]
@jdm should we merge this one already? That way I can start cleaning up the other one...

**Pull request merged**